### ISSUE TITLE
JAX-RS requires both the SSL and AppSecurity features in order to use a different key store

### ISSUE THREAD

**Issue by jkjome** [OpenLiberty/open-liberty#11103]
After asking a question on [StackOverflow: mpRestClient cert auth handshake failure under Open Liberty](https://stackoverflow.com/a/60385110/12924546) regarding encountering TLS handshake failures when using a custom keystore along with the `ssl-1.0` or `transportSecurity-1.0` features, it was determined by Andy McCright that using the `appSecurity-2.0` (or 3.0) feature would work around the issue, but should be considered a bug as only the SSL feature ought to be required to make use of a custom keystore.  Indeed, using the `appSecurity-20` feature did work around the TLS handshake failure issue, allowing the connection to succeed.  Andy asked me to report the bug, so I'm doing that here.

Related to this, note that I was encountering a `ClassNotFoundException` for `com.ibm.ws.jaxrs20.appsecurity.security.JaxRsSSLManager` when configuring the `liberty-maven-plugin` to reference `openliberty-runtime` in the `assemblyArtifact` config.  This resulted in continued TLS handshake failure when applying the suggested `appSecurity-2.0` workaround.  Only after changing the config to reference the `openliberty-microProfile3` package did the `ClassNotFoundException` finally go away and the TLS connection finally succeed.  Clearly the more complete Open Liberty microProfile3 package included the necessary SSL libraries by default, but it seems like Open Liberty should have auto-installed the necessary libraries if the base package (`openliberty-runtime`) didn't included them by default.  Maybe that's a distinct bug to address?

Below is a potentially related side-issue, so please let me know if this is part-and-parcel to the current issue or if it is a separate issue entirely?...

I also ran into downstream issue with hostname verification where, even though the TLS connection succeeds, hostname verification is applied by default and causes the connection to abort (because the hostname I call, and trust, doesn't exactly match what's in the the development-specific cert I'm using).  Based on the [Open Liberty 19.0.0.6 announcement](https://openliberty.io/blog/2019/06/21/microprofile-rest-client-19006.html#ssl), I thought that hostname verification was only added as of 19.0.0.6; with the [default being disabled](https://openliberty.io/docs/ref/config/ssl.html).  I tried including the `transportSecurity-1.0` feature and applying `verifyHostname="false"` to the `defaultSSLConfig` (along with a reference to my cutom keystore), but I found no change in behavior.  I even tried downgrading to Open Liberty 19.0.0.5 (presumably prior to inclusion of the hostname verification feature, based on the 19.0.0.6 announcement).  No matter what I do, hostname verification is always applied and always results in connection failure.

Please advise whether I'm just missing something or whether the hostname verification might be a separate bug from the main one here regarding needing `appSecurity-2.0` in order to use a custom keystore.

See the stacktrace below for the hostname verification issue (this one encountered under Open Liberty 19.0.0.5)...

```
Caused by: java.lang.RuntimeException: HostnameVerifier, socket reset for TTL
	at org.apache.cxf.transport.https.httpclient.DefaultHostnameVerifier.verify(DefaultHostnameVerifier.java:98)
	at java.base/sun.net.www.protocol.https.HttpsClient.checkURLSpoofing(HttpsClient.java:640)
	at java.base/sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:581)
	at java.base/sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:185)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1362)
	at java.base/sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1337)
	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:246)
	at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream.setupWrappedStream(URLConnectionHTTPConduit.java:302)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleHeadersTrustCaching(HTTPConduit.java:1391)
	at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.onFirstWrite(HTTPConduit.java:1352)
	at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream.onFirstWrite(URLConnectionHTTPConduit.java:335)
	at org.apache.cxf.io.AbstractWrappedOutputStream.write(AbstractWrappedOutputStream.java:47)
	at org.apache.cxf.io.AbstractThresholdOutputStream.unBuffer(AbstractThresholdOutputStream.java:89)
	at org.apache.cxf.io.AbstractThresholdOutputStream.write(AbstractThresholdOutputStream.java:63)
	at org.apache.cxf.io.AbstractWrappedOutputStream.write(AbstractWrappedOutputStream.java:51)
	at java.base/sun.nio.cs.StreamEncoder.writeBytes(StreamEncoder.java:233)
	at java.base/sun.nio.cs.StreamEncoder.implClose(StreamEncoder.java:337)
	at java.base/sun.nio.cs.StreamEncoder.close(StreamEncoder.java:161)
	at java.base/java.io.OutputStreamWriter.close(OutputStreamWriter.java:258)
	at org.glassfish.json.JsonGeneratorImpl.close(JsonGeneratorImpl.java:535)
	at org.eclipse.yasson.internal.Marshaller.marshall(Marshaller.java:80)
	at org.eclipse.yasson.internal.JsonBinding.toJson(JsonBinding.java:133)
	at com.ibm.ws.jaxrs21.providers.json.JsonBProvider.writeTo(JsonBProvider.java:172)
	at org.apache.cxf.jaxrs.utils.JAXRSUtils$3.run(JAXRSUtils.java:1466)
	at org.apache.cxf.jaxrs.utils.JAXRSUtils$3.run(JAXRSUtils.java:1463)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:734)
	at org.apache.cxf.jaxrs.utils.JAXRSUtils.writeMessageBody(JAXRSUtils.java:1463)
	at org.apache.cxf.jaxrs.client.AbstractClient.writeBody(AbstractClient.java:519)
	... 64 more
```

**Comment by 2ppisaHcl** [comment]
Related to mpRestClient-1.3 configuration, I noticed that in 19.0.0.12 the workaround (here I'm expecting that the ssl config in server.xml would be enough) for configuring programmatically the clients in terms of truststore and hostname verification, worked:
```
        RestClientBuilder builder = RestClientBuilder.newBuilder();
        builder.baseUrl(new URL("https://localhost:9490/service/with-client/test/"));
        builder.property("com.ibm.ws.jaxrs.client.ssl.config", "mySSLSettings");
        builder.property("com.ibm.ws.jaxrs.client.disableCNCheck", "true");
        builder.readTimeout(1, TimeUnit.MINUTES);
        builder.hostnameVerifier((hostname, session) -> true);

    <keyStore id="myKeyStore" location="key.jks" password="default" type="JKS"/>
    <keyStore id="myTrustStore" location="truststore.jks" password="default" type="JKS"/>

    <ssl id="mySSLSettings"
         keyStoreRef="myKeyStore"
         trustStoreRef="myTrustStore"
         sslProtocol="TLSv1.2"
         clientAuthenticationSupported="true"
         verifyHostname="false"
         performURLHostNameVerification="false"/>

    <sslDefault sslRef="mySSLSettings"/>

``` 
In 20.0.0.6 (I and maybe the previous ones) It also doesn't work anymore. 
The behaviour seems to be that when the line below is commented out        
``` 
builder.hostnameVerifier((hostname, session) -> true); 
``` 
the truststore configured in ssl tag element (or programmatically via "com.ibm.ws.jaxrs.client.ssl.config" property) is not used and apache cxf fallbacks to JMV default keystore and raises an 
**javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target**;
instead when the line hostnameVerifier() is not commented the error was
 java.lang.RuntimeException: java.lang.RuntimeException: HostnameVerifier, socket reset for TTL
(In reason of that I'm not confident that for the issue #11108 the right status is "Closed" or at least it does't works when
there is a custom configured ssl in server.xml with the need of skip the hostname verification).

**Comment by 2ppisaHcl** [comment]
An update, The programmatic microprofile rest client setup works iff I specify at jvm.options level
-Dcom.ibm.ws.jaxrs.client.security.overrideUserTLSConfig=true (As  stated in #11378 - I not found other doc about)

**Comment by andymc12** [comment]
I believe that the fix for this would require a break in zero-migration.
